model/src/main/kotlin/ScannerOptions.kt

```diff
@@ -0,0 +1,401 @@
+     */
+    inline fun <T : ScannerOption> checkOption(optionClass: Class<T>, test: (T) -> Boolean): Boolean =
+        getOption(optionClass)?.let { test(it) } ?: false
+
+    /**
+     * Execute a compatibility check on the option with the given type that depends on the [strict] mode flag.
+     * If the flag is *true*, an option of this type must be present and satisfy the [predicate][test]
+     * provided. Otherwise, an option with this type only needs to be present.
+     */
+    inline fun <reified T : ScannerOption> checkPresenceOrStrict(strict: Boolean, test: (T) -> Boolean): Boolean =
+        checkPresenceOrStrict(T::class.java, strict, test)
+
+    /**
+     * Execute a compatibility check on the option with the given [optionClass] that depends on the [strict] mode
+     * flag. If the flat is *true*, an option of this class must be present and satisfy the [predicate][test]
+     * provided. Otherwise, an option of this class only needs to be present.
+     */
+    inline fun <T : ScannerOption> checkPresenceOrStrict(
+        optionClass: Class<T>,
+        strict: Boolean,
+        test: (T) -> Boolean
+    ): Boolean =
+        (strict && checkOption(optionClass, test)) || (!strict && contains(optionClass))
+
+    /**
+     * Check whether scan results produced by a scanner run with the options stored in this instance are a subset of
+     * the results produced by the options in [other], taking the given [strict] flag into account. This function
+     * can be used to check whether a stored scan result with specific options can be used instead of running the
+     * scanner again. This basically means that the result contains at least the information the new scanner run
+     * would produce.
+     */
+    fun isSubsetOf(other: ScannerOptions, strict: Boolean): Boolean {
+        return options.all { it.isCompatible(other, strict) }
+    }
+}
+
+/**
+ * A base class to represent complex scanner options that can have themselves sub options. This class manages the
+ * sub options and implement the compatibility check based on them.
+ */
+sealed class ScannerOptionWithSubOptions<T : ScannerOptionWithSubOptions<T>>(
+    /**
+     * Stores the class of this option. This is required to access the correct counterpart from a [ScannerOptions]
+     * object during a compatibility check.
+     */
+    val optionClass: Class<T>
+) : ScannerOption() {
+
+    /** The sub options defined for this option. */
+    abstract val subOptions: SubOptions
+
+    /**
+     * Check whether this option is compatible with the given set of [options] doing either a strict or lenient
+     * comparison depending on the [strict] flag. This implementation checks whether the [options] contain an
+     * instance of this [optionClass] with equivalent [subOptions].
+     */
+    override fun isCompatible(options: ScannerOptions, strict: Boolean): Boolean =
+        options.checkPresenceOrStrict(optionClass, strict) {
```

Comment:
Couldn't you just use `this::class` here instead of making this class generic and storing the `optionClass`?

Reply:
Could you perhaps elaborate a bit? I am not so familiar with Kotlin's reflection capabilities. Is it then still possible to query specific options in a type-safe way?
Originally, I hoped to solve the access to specific sub option classes using reified types. But this was not working because sometimes dynamic classes need to be looked up.